app/src/main/java/com/example/sharingang/MapFragment.kt

```diff
@@ -120,42 +126,36 @@ class MapFragment : Fragment(), OnMapReadyCallback {
         )
     }
 
-    private fun moveLastLocationMarker() {
-        lastLocationMarker?.remove()
-        addLastLocationMarker()
-    }
+    private fun initMap() {
+        lifecycle.coroutineScope.launchWhenCreated {
+            map = binding.mapView.awaitMap()
 
-    private fun addLastLocationMarker() {
-        lastLocationMarker = map?.addMarker(
-            MarkerOptions().position(
-                LatLng(
-                    lastLocation!!.latitude,
-                    lastLocation!!.longitude
-                )
+            initCluster()
+
+            addItemMarkers(viewModel.searchResults.value ?: listOf())
+
+            doOrGetPermission(
+                this@MapFragment,
+                Manifest.permission.ACCESS_FINE_LOCATION,
+                { startLocationUpdates() },
+                requestPermissionLauncher
             )
-        )
+        }
     }
 
-    override fun onMapReady(googleMap: GoogleMap) {
-        map = googleMap
-        map?.setOnMarkerClickListener { marker: Marker ->
-            map?.animateCamera(CameraUpdateFactory.newLatLng(marker.position))
-            if (marker != lastLocationMarker) {
-                val markerItem = marker.tag as Item
-                this.findNavController().navigate(
-                    MapFragmentDirections.actionMapFragmentToDetailedItemFragment(markerItem)
-                )
-            }
+    private fun initCluster() {
+        markerManager = MarkerManager(map)
+        clusterManager = ClusterManager<MapItem>(context, map, markerManager)
+        clusterManager!!.renderer = ItemRenderer(context, map, clusterManager)
+        clusterManager!!.setOnClusterItemClickListener { marker ->
+            map!!.animateCamera(CameraUpdateFactory.newLatLng(marker.position))
+            findNavController().navigate(
+                MapFragmentDirections.actionMapFragmentToDetailedItemFragment(marker.item)
+            )
             true
         }
-        addItemMarkers(viewModel.searchResults.value ?: listOf())
 
-        doOrGetPermission(
-            this,
-            Manifest.permission.ACCESS_FINE_LOCATION,
-            { startLocationUpdates() },
-            requestPermissionLauncher
-        )
+        map!!.setOnCameraIdleListener(clusterManager)
```

Comment:
What does this do?

Reply:
`OnCameraIdle` is the event when the user finished moving the map. The cluster manager needs to receive this event so it can recompute the cluster (for example when the user zoom out, it would cluster them together)

Reply:
Thanks!